### PR TITLE
[AMORO-1720] Amoro Mixed Format KeyedTable support getting `optimizedSequence` from snapshot level

### LINKED ISSUE #1720: [Subtask]: Store max transaction id on the snapshot level for Mixed Format Table
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### Description

To support Time Travel (MOR on history snapshot) or optimize on history snapshot, we should store the max transaction id of each partition for Mixed Format KeyedTable on the snapshot level.

### Parent issue

#1354 

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangtaohz**
## Why are the changes needed?

resolve #1720 

## Brief change log

  - using `Puffin` file to store optimizedSequence 
  - remove the legacy optimizedSequence
  - the fromSequence and toSequence should not be -1 when there are no change files

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by zhoujinsong** [comment]
I would open another PR to change the project version to `0.6.0-SNAPSHOT`, so you can rollback the version change here.

File: core/src/main/java/com/netease/arctic/utils/PuffinUtil.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class PuffinUtil {
```
Comment (baiyangtx) [inline core/src/main/java/com/netease/arctic/utils/PuffinUtil.java:52]: What does this name mean? If it is copied from Iceberg, should we describe it in the comments?

I think some comments should be added to describe what this class is for. And comments should also be added to public methods.

File: core/src/main/java/com/netease/arctic/table/TableProperties.java [hunk 2]
```diff
@@ -33,6 +33,9 @@ public class TableProperties {
 
   private TableProperties() {
   }
+  
+  public static final String TABLE_VERSION = "table.version";
+  public static final String TABLE_VERSION_DEFAULT = "0.5.0";
```
Comment (baiyangtx) [inline core/src/main/java/com/netease/arctic/table/TableProperties.java:38]: Has this configuration been used before? Using the project version number as the table version doesn't seem like a good choice.

Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/table/TableProperties.java:38]: It's useless code, and I will remove that.

File: core/src/main/java/com/netease/arctic/utils/PuffinUtil.java (hunk 1 above)
Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/utils/PuffinUtil.java:52]: Good suggestion, I will add some comments. `Puffin` file is a definition in Iceberg. [[doc](https://iceberg.apache.org/puffin-spec/)]

**Comment by shidayang** [comment]
Title Arctic had renamed to Amoro

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/KeyedTableCommit.java [hunk 3]
```diff
@@ -75,8 +75,7 @@ public void commit() throws OptimizingCommitException {
     Set<DeleteFile> addedDeleteFiles = Sets.newHashSet();
     Set<DeleteFile> removedDeleteFiles = Sets.newHashSet();
 
-    StructLikeMap<Long> partitionOptimizedSequence =
-        TablePropertyUtil.getPartitionOptimizedSequence(table.asKeyedTable());
+    StructLikeMap<Long> partitionOptimizedSequence = PuffinUtil.reader(table.asKeyedTable()).readOptimizedSequence();
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/KeyedTableCommit.java:78]: Should we can wrap PuffinUtil.reader(table.asKeyedTable()).readOptimizedSequence() logic into KeyedTable interface?

Comment (wangtaohz) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/KeyedTableCommit.java]: I think it is not suitable to include this in the `KeyedTable` interface. We do not want to expose the `optimized sequence` to the users of the `Table` API, even though the current encapsulation is not yet perfect.

BTW, I have move the `PuffinUtil.reader(table.asKeyedTable()).readOptimizedSequence()` to `com.netease.arctic.utils.ArcticTableUtil#readOptimizedSequence`.

**Comment by wangtaohz** [comment]
@zhoujinsong I make some refactoring to `PuffinUtil`, and now its name has changed to `StatisticsFileUtil`. You could take another look at your convenience.

File: core/src/main/java/com/netease/arctic/utils/ArcticTableUtil.java [hunk 4]
```diff
@@ -34,4 +46,79 @@ public static String tableRootLocation(ArcticTable arcticTable) {
...
+      }
+      return result.get(0);
+    } else {
+      return null;
+    }
+  }
+
+  private static StructLikeMap<Long> readLegacyPartitionProperties(
+      UnkeyedTable table, String type) {
+    // to be compatible with old Amoro version 0.5.0 which didn't use puffin file and stored the
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/utils/ArcticTableUtil.java:110]: Nit: Comments should start with a capital letter.

Comment (zhoujinsong) [reply core/src/main/java/com/netease/arctic/utils/ArcticTableUtil.java:110]: Spotless automatically wraps lines at 100 characters, but you can make some adjustments for better readability here. Additionally, you can adjust the line length in your local IDEA to match the behavior of Spotless.

**Review by zhoujinsong — COMMENTED** [review]
@wangtaohz 
This PR look good to me. I left some small suggestion, please take a look.

File: core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/** Util class for write and read Iceberg statistics file {@link StatisticsFile}. */
+public class StatisticsFileUtil {
+
+  public static Writer writer(Table table, long snapshotId, long sequenceNumber) {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java:58]: Perhaps opening a separate method to set the written snapshot would be more flexible?

File: core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    private boolean closed = false;
+
+    private Writer(Table table, long snapshotId, long sequenceNumber) {
+      this.snapshotId = snapshotId;
+      this.sequenceNumber = sequenceNumber;
+      this.outputFile =
+          table
+              .io()
+              .newOutputFile(
+                  table.location() + "/data/" + snapshotId + "-" + UUID.randomUUID() + ".puffin");
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java:80]: Write Puffin files under `{table_location}/data/puffin` path for better management.

File: core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+  /**
+   * Find the statistics file that satisfies the condition and belongs to the latest snapshot.
+   *
+   * @param table - Iceberg Table
+   * @param currentSnapshotId - find from this snapshot to its ancestors
+   * @param condition - the condition to filter the statistics file
+   * @return the statistics file, return null if not found
+   */
+  public static List<StatisticsFile> findLatestValidStatisticsFiles(
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java:187]: Most of the time we need to find the static file belonging to a certain snapshot to ensure the correctness of the logic. Searching for its parent's static file only makes sense in some special scenarios when it cannot be found.

File: core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java (hunk 5 above)
Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java]: I add a WriterBuild to support it.

File: core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java (hunk 7 above)
Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/utils/StatisticsFileUtil.java]: I introduce a mark in snapshot summary in order to find the snapshot.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**